src/rust/jsg/ffi.c++

```diff
@@ -0,0 +1,263 @@
+  return to_ffi(args->This());
+}
+
+size_t fci_get_length(FunctionCallbackInfo* args) {
+  return args->Length();
+}
+
+Local fci_get_arg(FunctionCallbackInfo* args, size_t index) {
+  return to_ffi((*args)[index]);
+}
+
+void fci_set_return_value(FunctionCallbackInfo* args, Local value) {
+  args->GetReturnValue().Set(local_from_ffi<v8::Value>(kj::mv(value)));
+}
+
+Global create_resource_template(v8::Isolate* isolate, const ResourceDescriptor& descriptor) {
+  // Construct lazily.
+  v8::EscapableHandleScope scope(isolate);
+
+  v8::Local<v8::FunctionTemplate> constructor;
+  KJ_IF_SOME(descriptor, descriptor.constructor) {
+    constructor = v8::FunctionTemplate::New(isolate,
+        reinterpret_cast<v8::FunctionCallback>(reinterpret_cast<void*>(descriptor.callback)));
+  } else {
+    constructor = v8::FunctionTemplate::New(isolate, &workerd::jsg::throwIllegalConstructor);
+  }
+
+  auto prototype = constructor->PrototypeTemplate();
+
+  // Signatures protect our methods from being invoked with the wrong `this`.
+  auto signature = v8::Signature::New(isolate, constructor);
+
+  auto instance = constructor->InstanceTemplate();
+
+  instance->SetInternalFieldCount(workerd::jsg::Wrappable::INTERNAL_FIELD_COUNT);
+
+  auto classname = ::workerd::jsg::check(v8::String::NewFromUtf8(
+      isolate, descriptor.name.data(), v8::NewStringType::kNormal, descriptor.name.size()));
+
+  if (workerd::jsg::getShouldSetToStringTag(isolate)) {
+    prototype->Set(v8::Symbol::GetToStringTag(isolate), classname, v8::PropertyAttribute::DontEnum);
+  }
+
+  // Previously, miniflare would use the lack of a Symbol.toStringTag on a class to
+  // detect a type that came from the runtime. That's obviously a bit problematic because
+  // Symbol.toStringTag is required for full compliance on standard web platform APIs.
+  // To help use cases where it is necessary to detect if a class is a runtime class, we
+  // will add a special symbol to the prototype of the class to indicate. Note that
+  // because this uses the global symbol registry user code could still mark their own
+  // classes with this symbol but that's unlikely to be a problem in any practical case.
+  auto internalMarker =
+      v8::Symbol::For(isolate, ::workerd::jsg::v8StrIntern(isolate, "cloudflare:internal-class"));
+  prototype->Set(internalMarker, internalMarker,
+      static_cast<v8::PropertyAttribute>(v8::PropertyAttribute::DontEnum |
+          v8::PropertyAttribute::DontDelete | v8::PropertyAttribute::ReadOnly));
+
+  constructor->SetClassName(classname);
+
+  // auto& typeWrapper = static_cast<TypeWrapper&>(*this);
+
+  // ResourceTypeBuilder<TypeWrapper, T, isContext> builder(
+  //     typeWrapper, isolate, constructor, instance, prototype, signature);
+
+  // if constexpr (isDetected<GetConfiguration, T>()) {
+  //   T::template registerMembers<decltype(builder), T>(builder, configuration);
+  // } else {
+  //   T::template registerMembers<decltype(builder), T>(builder);
+  // }
```

Comment:
Does this need a TODO or similar?

Reply:
Yes. This will be done on a separate pr that adds type generation support.